### PR TITLE
[3875] Refactor the tree event classes related to the explorer

### LINKED ISSUE #3875: Refactor the tree event classes related to the explorer

### PR DESCRIPTION
**Pull request by jerome-obeo**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3875

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: CHANGELOG.adoc
Comment (sbegaudeau) [inline CHANGELOG.adoc]: sirius-components-trees

File: packages/forms/frontend/sirius-components-widget-reference/src/components/ModelBrowserTreeView.tsx
Comment (sbegaudeau) [inline packages/forms/frontend/sirius-components-widget-reference/src/components/ModelBrowserTreeView.tsx]: I suspect that this make the reference widget unusable without the presence of the explorer. There are applications that are integrating Sirius Components without Sirius Web. Sirius Components cannot make any assumptions on the presence or capabilities of Sirius Web. This will have to change in favor of reference-widget specific capabilities.

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/ExplorerEventProcessorFactory.java
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/ExplorerEventProcessorFactory.java]: Given that we now explicitly what we are looking for since we are not in some generic tree-related code but instead in some explorer related code, could you instead search for the description with the id `ExplorerDescriptionProvider#DESCRIPTION_ID`.

Thanks to that:
- you don't need to have this TREE_ID variable here
- you don't need to have a special canCreatPredicate in ExplorerDescriptionProvider
- you can remove all usages of this TREE_ID
- you don't have to introduce the useless interface ITreeConfiguration to hold this variable

And you will have to provide the concept necessary for the reference widget explicitly in its collaborative project.

File: packages/sirius-web/backend/sirius-web-application/src/main/resources/schema/explorer.graphqls
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/resources/schema/explorer.graphqls:7]: Just as an information, `treeId` is probably what will be replaced in favor of `treeDescriptionId` to let the frontend specify to the backend which tree description to use to compute the explorer later.

I suspect that the treeId variable could be fully removed in this PR without any impact, it is now a duplicate of the information contained in the fact that we are using a GraphQL field named `explorerEvent`.

File: packages/sirius-web/backend/sirius-web-tests/src/main/java/org/eclipse/sirius/web/tests/services/explorer/ExplorerEventSubscriptionRunner.java
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-tests/src/main/java/org/eclipse/sirius/web/tests/services/explorer/ExplorerEventSubscriptionRunner.java:33]: TREE_EVENT_SUBSCRIPTION -> EXPLORER_EVENT_SUBSCRIPTION on the line 32

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/ModelBrowserExpandAllControllerTests.java
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/ModelBrowserExpandAllControllerTests.java]: This should make you stop because such dependency between the reference widget from the form representation and the explorer cannot be introduced.

File: packages/trees/frontend/sirius-components-trees/src/views/TreeViewMachine.ts
Comment (sbegaudeau) [inline packages/trees/frontend/sirius-components-trees/src/views/TreeViewMachine.ts]: That's a bit odd and I suspect that it will just become more and more untenable (try fixing the issue with the reference widget with this code and I suspect that it won't be possible without adding odds dependencies). I suspect that we will have to move to the same situation as in forms with the subscription moved outside of TreeView with a dedicated useExplorerEvent subscription and a TreeView component which would simply receive the payload.

File: packages/trees/frontend/sirius-components-trees/src/views/getTreeEventSubscription.ts
Comment (sbegaudeau) [inline packages/trees/frontend/sirius-components-trees/src/views/getTreeEventSubscription.ts:13]: From what I've seen in the code, if I'm right, you will have to duplicate this subscription to have a getExplorerEventSubscription in `sirius-web-application` instead and just keep the function `recursiveGetChildren` and the fragment here...

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
It is absolutely critical that the reference widget keeps being usable without the explorer. As such, it will be necessary to have a modelBrowserEvent subscription with its event processor factory.

**Comment by sbegaudeau** [comment]
You should also keep an eye on [this PR](https://github.com/eclipse-sirius/sirius-web/pull/3870) by @florianbarbin since it will introduce yet another usage of the tree representation. I don't know yet which PR will be merged first but the other one will have to be updated to take into account these changes.

On the same topic, [the previous PR from Florian Barbin](https://github.com/eclipse-sirius/sirius-web/commit/5f845ea6cf68b24a1eb3977925df0fb36536f232) introduced another basic version of a TreeDescription in the view DSL in order to leverage it in the selection dialog. One of the goal of the future tree description that you will add in the view DSL is to replace both this one dedicated to the selection dialog (which is on purpose very minimal) and the one powering the form representation.

File: CHANGELOG.adoc
Comment (jerome-obeo) [reply CHANGELOG.adoc]: Done.

File: packages/sirius-web/backend/sirius-web-tests/src/main/java/org/eclipse/sirius/web/tests/services/explorer/ExplorerEventSubscriptionRunner.java
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-tests/src/main/java/org/eclipse/sirius/web/tests/services/explorer/ExplorerEventSubscriptionRunner.java:33]: Done

File: packages/forms/frontend/sirius-components-widget-reference/src/components/ModelBrowserTreeView.tsx
Comment (jerome-obeo) [reply packages/forms/frontend/sirius-components-widget-reference/src/components/ModelBrowserTreeView.tsx]: Exactly, ModelBrowser has its own `modelBrowserEvent` subscription now.

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/ExplorerEventProcessorFactory.java
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/views/explorer/ExplorerEventProcessorFactory.java]: You are right, I got rid of this variable.

File: packages/sirius-web/backend/sirius-web-application/src/main/resources/schema/explorer.graphqls
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web-application/src/main/resources/schema/explorer.graphqls:7]: Unchanged for the moment.
You are right this field will be renamed `treeDescriptionId` in the future task that tackles the explorer selection.
This ID will be sent from the frontend as the current description to build.

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/ModelBrowserExpandAllControllerTests.java
Comment (jerome-obeo) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/trees/ModelBrowserExpandAllControllerTests.java]: Right, there is a dedicated `ModelBrowserEventSubscritionRunner` helper class.
The test has been updated to use model browser related classes and not explorer one (EventInput for instance).

File: packages/trees/frontend/sirius-components-trees/src/views/TreeViewMachine.ts
Comment (jerome-obeo) [reply packages/trees/frontend/sirius-components-trees/src/views/TreeViewMachine.ts]: Yes, the subscription has been moved outside the TreeView component. This is now the parent component of TreeView, which is in charge of managing the subscription. For instance, ModelBrowserTreeView retrieves the tree through the subscription and passes it to its inner TreeView via props.

File: packages/trees/frontend/sirius-components-trees/src/views/getTreeEventSubscription.ts
Comment (jerome-obeo) [reply packages/trees/frontend/sirius-components-trees/src/views/getTreeEventSubscription.ts:13]: Done.

**Comment by jerome-obeo** [comment]
PR updated.
A new commit contains changes responding to comments.

**Review by frouene — APPROVED** [review]

File: packages/trees/frontend/sirius-components-trees/src/views/TreeView.tsx
Comment (sbegaudeau) [inline packages/trees/frontend/sirius-components-trees/src/views/TreeView.tsx]: TreeViewState to match our naming conventions

File: packages/trees/frontend/sirius-components-trees/src/views/TreeView.types.ts
Comment (sbegaudeau) [inline packages/trees/frontend/sirius-components-trees/src/views/TreeView.types.ts:21]: The TreeView component cannot be used directly as a workbench view anymore, it does not have to keep extending WorkbenchViewComponentProps it doesn't make sense. ExplorerView needs to do it (which it does) but not TreeView.

File: packages/trees/frontend/sirius-components-trees/src/views/TreeView.types.ts
Comment (sbegaudeau) [inline packages/trees/frontend/sirius-components-trees/src/views/TreeView.types.ts]: These are not the concern of sirius-components-trees it's not used in this npm package.

By putting this code here, you are trying to avoid having two almost identical copies of those types in `useModelBrowserSubscription` and `useExplorerSubscription`. Unless we have some maintainability issue with the code, some duplicated code is not an issue I care about. On the other hand, useless code (for `sirius-components-trees`) and useless coupling are both of the most critical issues that I'm trying to avoid.

Removing duplicated code always comes with one critical tradeoff, creating additional coupling.

Now both the lifecycle the explorer and the model browser are tied together with those types. What if someone wants to change one of those? Their first instinct will be to update `UseTreeSubscriptionState` and we will end up with useless code for the other one.

The state of `useExplorerSubscription` needs to fulfill the expectations of `useExplorersubscription` and the state of  `useModelBrowserSubscription` needs to fulfill the expectations of `useModelBrowserSubscription`. It just happens that today both have the same expectations.

File: packages/forms/frontend/sirius-components-widget-reference/src/components/ModelBrowserTreeView.tsx
Comment (sbegaudeau) [inline packages/forms/frontend/sirius-components-widget-reference/src/components/ModelBrowserTreeView.tsx]: Don't use this way for conditional rendering, be explicit because the implicit casting in JavaScript for boolean is funky

File: packages/trees/frontend/sirius-components-trees/src/views/TreeView.tsx
Comment (sbegaudeau) [inline packages/trees/frontend/sirius-components-trees/src/views/TreeView.tsx]: - Do not use this way to express conditional rendering
- Why does tree can be `null` when it is explicitly described as non null in its type?

File: packages/sirius-web/frontend/sirius-web-application/src/views/explorer/ExplorerView.tsx
Comment (sbegaudeau) [inline packages/sirius-web/frontend/sirius-web-application/src/views/explorer/ExplorerView.tsx]: Here you are giving a `GQLTree | null` to TreeView which expects a `GQLTree`. On the other hand, in `ModelBrowserTree` you are not rendering `TreeView` unless you have a `GQLTree`. So who is responsible for testing that `tree` is `null`?

Given the fact that both `ModelBrowser` is testing for `null` and the type of `tree` in `TreeViewProps` is `GQLTree`, I suspect that `ExplorerView` should do the job.

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I'll update the PR with fixes for my latest comments. After that I'll rebase it and merge it. No need to touch anything now

**Review by sbegaudeau — APPROVED** [review]
I'll wait for the build to merge it

**Pull request merged**
